File: infra/src/main/java/com/beat/infra/persistence/promotion/entity/PromotionJpaEntity.java
package com.beat.infra.persistence.promotion.entity;

import com.beat.domain.promotion.domain.CarouselNumber;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity(name = "Promotion")
@Table(name = "promotion")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class PromotionJpaEntity {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(nullable = false)
	private String promotionPhoto;

	@Column(name = "performance_id", nullable = true)
	private Long performanceId;

	@Column(nullable = false)
	private String redirectUrl;

	@Column(nullable = false)
	private boolean isExternal;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private CarouselNumber carouselNumber;

	private PromotionJpaEntity(Long id, String promotionPhoto, Long performanceId, String redirectUrl,
		boolean isExternal, CarouselNumber carouselNumber) {
		this.id = id;
		this.promotionPhoto = promotionPhoto;
		this.performanceId = performanceId;
		this.redirectUrl = redirectUrl;
		this.isExternal = isExternal;
		this.carouselNumber = carouselNumber;
	}

	public static PromotionJpaEntity rehydrate(Long id, String promotionPhoto, Long performanceId, String redirectUrl,
		boolean isExternal, CarouselNumber carouselNumber) {
		return new PromotionJpaEntity(id, promotionPhoto, performanceId, redirectUrl, isExternal, carouselNumber);
	}
}

File: infra/src/main/kotlin/com/beat/infra/persistence/promotion/entity/PromotionJpaEntity.kt
package com.beat.infra.persistence.promotion.entity

import com.beat.domain.promotion.domain.CarouselNumber
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table

@Entity(name = "Promotion")
@Table(name = "promotion")
class PromotionJpaEntity private constructor(
    id: Long?,
    promotionPhoto: String,
    performanceId: Long?,
    redirectUrl: String,
    isExternal: Boolean,
    carouselNumber: CarouselNumber,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(updatable = false)
    var id: Long? = id
        protected set

    @Column(nullable = false)
    var promotionPhoto: String = promotionPhoto
        protected set

    @Column(name = "performance_id", nullable = true)
    var performanceId: Long? = performanceId
        protected set

    @Column(nullable = false)
    var redirectUrl: String = redirectUrl
        protected set

    @Column(nullable = false)
    var isExternal: Boolean = isExternal
        protected set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var carouselNumber: CarouselNumber = carouselNumber
        protected set

    companion object {
        @JvmStatic
        fun rehydrate(
            id: Long?,
            promotionPhoto: String,
            performanceId: Long?,
            redirectUrl: String,
            isExternal: Boolean,
            carouselNumber: CarouselNumber,
        ): PromotionJpaEntity = PromotionJpaEntity(
            id = id,
            promotionPhoto = promotionPhoto,
            performanceId = performanceId,
            redirectUrl = redirectUrl,
            isExternal = isExternal,
            carouselNumber = carouselNumber,
        )
    }
}
